tracer/src/Datadog.Trace/Agent/StatsAggregator.cs

```diff
@@ -102,11 +108,33 @@ public bool AddRange(ArraySegment<Span> spans)
             {
                 for (int i = 0; i < spans.Count; i++)
                 {
-                    forceKeep |= AddToBuffer(spans.Array[i + spans.Offset]);
+                    AddToBuffer(spans.Array[i + spans.Offset]);
                 }
             }
+        }
+
+        public bool RunSamplers(ArraySegment<Span> trace)
+        {
+            // TODO: Add ability to disable the RareSampler, which the trace agent has
+            // Run the RareSampler early to make sure the signature gets counted
```

Comment:
> The sampling part would be called only after all other samplers have been called.

So after a bit of reading the agent code and testing, this isn't quite true. Here's the algorithm we need:
- Run the RareSampler first. If the sampling priority is > 0, the RareSampler will not mark any spans with "_dd.rare = 1". Otherwise, the first rare span in the chunk that's determined to be rare is marked as "_dd.rare = 1"
- If no other samplers decide to sample the trace chunk, return the result of the RareSampler as the decision for the trace

Reply:
In commit 69b31153d4033a5c64cf73504aa3dff90bf6a95d I updated the RareSampler slightly. I documented the reasons for the algorithm so the algorithm itself didn't change too much, but I tried to refactor and add comments so it's clearer. Let me know what you think

Reply:
> Run the RareSampler first. If the sampling priority is > 0, the RareSampler will not mark any spans with "_dd.rare = 1". Otherwise, the first rare span in the chunk that's determined to be rare is marked as "_dd.rare = 1"
If no other samplers decide to sample the trace chunk, return the result of the RareSampler as the decision for the trace

So to make sure IUC, the difference in behaviour with what I suggested is the fact that a span can be marked as `_dd.rare = 1` even though it is sampled by another sampler (eg we have a span with prio <=0, it is rare, we add the tag, but it also contains an error, so it is returned anyway). Whereas in my suggestion, we wouldn't add the `_dd.rare = 1` in that case. Right?